File: solutions/217.Contains_Duplicate/AC_map_nlogn.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_map_nlogn.cpp
*  Create Date: 2015-07-13 21:39:31
*  Descripton:   
*/


#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        map<int, int> mp;
        for (auto i : nums) {
            if (mp.find(i) != mp.end()) {
                return true;
            }
            mp[i] = 1;
        }
        return false;
    }
};

int main() {

    return 0;
}

File: solutions/290.Word_Pattern/AC_map_nlogn.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_map_nlogn.py
# Create Date: 2015-10-24 23:02:14
# Usage:       AC_map_nlogn.py
# Descripton:

class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        mp = ["" for _ in xrange(26)]
        strs = str.split()
        if len(strs) != len(pattern):
            return False
        for i, c in enumerate(pattern):
            idx = ord(c) - ord('a')
            if mp[idx] == "":
                for v in mp:
                    if v == strs[i]:
                        return False
                mp[idx] = strs[i]
            else:
                if mp[idx] != strs[i]:
                    return False
        return True

s = Solution()
print(s.wordPattern("abba", "d c c d"))
